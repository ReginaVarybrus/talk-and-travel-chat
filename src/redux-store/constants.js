const ULRs = {
  register: 'authentication/register',
  login: 'authentication/login',
  logout: 'authentication/logout',
  currentUser: 'current',
  updateUser: 'users',
  userInfo: userId => `users/${userId}`,
  getMainCountryChatByName: countryName => `chats/${countryName}/main`,
  userCountries: userId => `countries/user/${userId}`,
  subscriptionToMessages: chatId => `/countries/${chatId}/messages`,
  subscriptionToUserErrors: userId => `/user/${userId}/errors`,
  getChatsParticipants: chatId => `/chats/${chatId}/users`,
  joinToGroupChat: '/chat/events.joinChat',
  leaveOutGroupChat: '/chat/events.leaveChat',
  startTyping: '/chat/events.startTyping',
  stopTyping: '/chat/events.stopTyping',
  getPrivateChats: userId => `chats/user/${userId}/private`,
  createPrivateChat: 'chats/private',
  getChatsMessages: chatId => `chats/${chatId}`,
};

export default ULRs;
