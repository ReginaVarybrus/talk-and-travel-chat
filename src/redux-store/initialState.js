const initialState = {
  token: null,
  userDto: {
    userName: '',
    userEmail: '',
    avatar: '',
  },
  name: '',
  flagCode: '',
  userId: '',
  countryRooms: [],

  isLoggedIn: false,
  isRefresh: true,
  error: null,
};

export default initialState;
