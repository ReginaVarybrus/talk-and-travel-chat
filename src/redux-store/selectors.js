export const getIsLoggedIn = state => state.auth.isLoggedIn;
export const getUserRefresh = state => state.auth.isRefresh;
export const getUser = state => state.user;
