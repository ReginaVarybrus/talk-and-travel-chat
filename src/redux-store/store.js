import { configureStore, combineSlices } from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { authSlice } from './slices/authSlice';
import { userSlice } from './slices/userSlice';

const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['token'],
};

const rootPersistConfig = {
  key: 'root',
  storage,
  blacklist: ['auth'],
};

const rootReducer = combineSlices({
  auth: persistReducer(authPersistConfig, authSlice.reducer),
  user: userSlice.reducer,
});

const persistedReducer = persistReducer(rootPersistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);
