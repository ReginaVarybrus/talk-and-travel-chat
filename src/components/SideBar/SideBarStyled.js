import styled from 'styled-components';
import { device } from '@/constants/mediaQueries';
import { LuLogOut, LuMessagesSquare } from 'react-icons/lu';
import { TbUser, TbUsers } from 'react-icons/tb';

export const iconProperties = (
  width = '24px',
  height = '48px',
  marginRight = '0'
) => `
  width: ${width};
  height: ${height};
  margin-right: ${marginRight};
  stroke: var(--color-grey-9);
`;

export const SideBarStyled = styled.div`
  display: none;
  @media ${device.tablet} {
    display: flex;
    flex-direction: column;
    align-items: center;
    height: 100vh;
    background: var(--color-blue-1);
  }
`;

export const ButtonsFrame = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
  margin: 48px 0;
`;

export const Text = styled.p`
  line-height: 20px;
  letter-spacing: -0.011em;
  color: var(--color-grey-9);
`;

export const UserIcon = styled(TbUser)`
  ${iconProperties()};
`;

export const RoomsIcon = styled(TbUsers)`
  ${iconProperties('24px', '24px', '12px')};
  @media ${device.tablet} {
    ${iconProperties()};
  }
`;

export const DMsIcon = styled(LuMessagesSquare)`
  ${iconProperties('24px', '24px', '12px')};
  @media ${device.tablet} {
    ${iconProperties()};
  }
`;

export const LogoutIcon = styled(LuLogOut)`
  ${iconProperties()};
`;

export const SideBarButton = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 56px;
  height: 76px;
  cursor: pointer;
  color: ${({ $isActive }) =>
    $isActive ? 'var(--color-brand-blue)' : 'var(--color-grey-9)'};
  transition: color 0.3s;

  ${Text} {
    color: currentColor;
    font-weight: ${({ $isActive }) => $isActive && 'bolder'};
  }

  svg {
    stroke: currentColor;
  }

  &:hover {
    color: var(--color-blue-3);

    svg {
      stroke: currentColor;
    }
  }
`;

export const ChatsButtonsFrame = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 51px;
  height: 166px;
`;
