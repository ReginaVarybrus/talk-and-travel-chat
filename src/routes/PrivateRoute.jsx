import { Navigate } from 'react-router-dom';
import { useSelector } from 'react-redux';
import {
  getIsLoggedIn,
  getUserRefresh,
} from '@/redux-store/AuthOperations/selectors';
import { routesPath } from '@/routes/routesConfig';

export default function PrivateRoute({
  component: Component,
  redirectTo = routesPath.MAIN,
}) {
  const isLoggedIn = useSelector(getIsLoggedIn);
  const isRefresh = useSelector(getUserRefresh);

  const shouldRedirect = !isLoggedIn && !isRefresh;

  return shouldRedirect ? <Navigate to={redirectTo} /> : <Component />;
}
