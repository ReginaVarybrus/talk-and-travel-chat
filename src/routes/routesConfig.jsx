import { lazy } from 'react';
import PrivateRoute from '@/routes/PrivateRoute';
import { createBrowserRouter } from 'react-router-dom';

const ROOT = import.meta.env.BASE_URL;

export const routesPath = {
  MAIN: ROOT,
  LOGIN: `${ROOT}login`,
  REGISTER: `${ROOT}register`,
  APP: `${ROOT}app`,
  CHAT: `${ROOT}app/chat/`,
  ACCOUNT: `${ROOT}app/account/`,
};

const importComponent = {
  MAIN: lazy(() => import('@/routes/MainRoute/MainRoute')),
  LOGIN: lazy(() => import('@/routes/LoginRoute/LoginRoute')),
  LAYOUT: lazy(() => import('@/components/Layout/Layout')),
  REGISTER: lazy(() => import('@/routes/RegisterRoute/RegisterRoute')),
  CHAT: lazy(() => import('@/routes/ChatRoute/ChatRoute')),
  ACCOUNT: lazy(() => import('@/routes/AccountRoute/AccountRoute')),
  ERROR_COMPONENT: lazy(() => import('@/routes/Page404/Page404')),
};

export const router = createBrowserRouter([
  {
    path: routesPath.MAIN,
    Component: importComponent.MAIN,
    errorElement: importComponent.ERROR_COMPONENT,
  },
  {
    path: routesPath.LOGIN,
    Component: importComponent.LOGIN,
  },
  {
    path: routesPath.REGISTER,
    Component: importComponent.REGISTER,
  },
  {
    path: routesPath.APP,
    element: (
      <PrivateRoute
        component={importComponent.LAYOUT}
        redirectTo={routesPath.LOGIN}
      />
    ),
    children: [
      {
        path: routesPath.CHAT,
        element: (
          <PrivateRoute
            component={importComponent.CHAT}
            redirectTo={routesPath.LOGIN}
          />
        ),
      },
      {
        path: routesPath.ACCOUNT,
        element: (
          <PrivateRoute
            component={importComponent.ACCOUNT}
            redirectTo={routesPath.LOGIN}
          />
        ),
      },
    ],
  },
]);
