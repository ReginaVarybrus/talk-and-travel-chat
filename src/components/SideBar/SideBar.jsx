import { useNavigate, useLocation } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { routesPath } from '@/routes/routesConfig';
import { getUser } from '@/redux-store/selectors';
import { logOut } from '@/redux-store/AuthOperations/AuthOperations';
import {
  SideBarStyled,
  ButtonsFrame,
  SideBarButton,
  ChatsButtonsFrame,
  Text,
  UserIcon,
  RoomsIcon,
  DMsIcon,
  LogoutIcon,
} from './SideBarStyled';

const SideBar = () => {
  const { userName } = useSelector(getUser) || {};
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const location = useLocation();
  const currentPage = location.pathname;

  const handleProfileOpen = () => {
    navigate(routesPath.ACCOUNT);
  };

  const handleRoomsOpen = () => {
    navigate(routesPath.ROOMS);
  };

  const handleDMsOpen = () => {
    navigate(routesPath.DMS);
  };

  const handleLogOut = (event, values) => {
    dispatch(logOut(values));
    navigate(routesPath.MAIN);
  };

  return (
    <SideBarStyled>
      <ButtonsFrame>
        <SideBarButton
          onClick={handleProfileOpen}
          $isActive={currentPage === routesPath.ACCOUNT}
        >
          <UserIcon />
          <Text>{userName}</Text>
        </SideBarButton>
        <ChatsButtonsFrame>
          <SideBarButton
            onClick={handleRoomsOpen}
            $isActive={currentPage === routesPath.ROOMS}
          >
            <RoomsIcon />
            <Text>Rooms</Text>
          </SideBarButton>
          <SideBarButton
            onClick={handleDMsOpen}
            $isActive={currentPage === routesPath.DMS}
          >
            <DMsIcon />
            <Text>DMs</Text>
          </SideBarButton>
        </ChatsButtonsFrame>
        <SideBarButton onClick={handleLogOut}>
          <LogoutIcon />
          <Text>LogOut</Text>
        </SideBarButton>
      </ButtonsFrame>
    </SideBarStyled>
  );
};

export default SideBar;
