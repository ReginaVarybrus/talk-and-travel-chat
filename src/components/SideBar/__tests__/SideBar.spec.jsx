import { describe, expect } from 'vitest';
import { render } from '@testing-library/react';

import { Provider } from 'react-redux';
import { RouterProvider } from 'react-router-dom';
import { store } from '@/redux-store/store';
import { router } from '@/routes/routesConfig';

import SideBar from '../SideBar';

const renderComponent = children => {
  return render(
    <Provider store={store}>
      <RouterProvider router={router}>{children}</RouterProvider>
    </Provider>
  );
};

describe('SideBar component test', () => {
  it('render SideBar component', () => {
    const { sideBar } = renderComponent(<SideBar />);

    expect(sideBar).toMatchSnapshot();
  });
});
