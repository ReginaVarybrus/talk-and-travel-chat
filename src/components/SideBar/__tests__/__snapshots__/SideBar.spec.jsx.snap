// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideBar component test > render SideBar component 1`] = `undefined`;

exports[`SideBar component test > renders SideBar component 1`] = `undefined`;
